File: packages/csharp/src/AlphaTab/Core/EcmaScript/DataView.cs
using System;

namespace AlphaTab.Core.EcmaScript;

internal class DataView
{
    public ArrayBuffer Buffer { get; }
    public double ByteOffset { get; }
    public double ByteLength { get; }


    public DataView(ArrayBuffer buffer)
    {
        Buffer = buffer;
    }

    public DataView(ArrayBuffer buffer, double byteOffset, double byteLength)
    {
        Buffer = buffer;
        ByteOffset = byteOffset;
        ByteLength = byteLength;
    }

    public double GetInt16(double offset, bool littleEndian)
    {
        if (littleEndian == BitConverter.IsLittleEndian)
        {
            return BitConverter.ToInt16(Buffer.Raw, (int)(ByteOffset + offset));
        }

        var bytes = new byte[sizeof(short)];
        System.Buffer.BlockCopy(Buffer.Raw, (int)(ByteOffset + offset), bytes, 0,
            bytes.Length);
        System.Array.Reverse(bytes);
        return BitConverter.ToInt16(bytes, 0);
    }

    public double GetFloat32(double offset, bool littleEndian = true)
    {
        if (littleEndian == BitConverter.IsLittleEndian)
        {
            return BitConverter.ToSingle(Buffer.Raw, (int)(ByteOffset + offset));
        }

        var bytes = new byte[sizeof(float)];
        System.Buffer.BlockCopy(Buffer.Raw, (int)(ByteOffset + offset), bytes, 0,
            bytes.Length);
        System.Array.Reverse(bytes);
        return BitConverter.ToSingle(bytes, 0);
    }
}

File: packages/kotlin/src/android/src/main/java/alphaTab/core/ecmaScript/DataView.kt
package alphaTab.core.ecmaScript

import java.nio.ByteOrder

@ExperimentalUnsignedTypes
internal class DataView {
    private val _littleEndianBufferView: java.nio.ByteBuffer
    private val _bigEndianBufferView: java.nio.ByteBuffer

    val buffer: ArrayBuffer
    val byteOffset: Double
    val byteLength: Double

    constructor(buffer: ArrayBuffer) {
        this.buffer = buffer
        byteOffset = 0.0
        byteLength = buffer.size.toDouble()
        this._littleEndianBufferView =
            java.nio.ByteBuffer.wrap(buffer.asByteArray()).order(ByteOrder.LITTLE_ENDIAN)
        this._bigEndianBufferView =
            java.nio.ByteBuffer.wrap(buffer.asByteArray()).order(ByteOrder.BIG_ENDIAN)

    }

    constructor(buffer: ArrayBuffer, byteOffset: Double, byteLength: Double) {
        this.buffer = buffer
        this.byteOffset = byteOffset
        this.byteLength = byteLength
        this._littleEndianBufferView =
            java.nio.ByteBuffer.wrap(buffer.asByteArray(), byteOffset.toInt(), byteLength.toInt())
                .order(ByteOrder.LITTLE_ENDIAN)
        this._bigEndianBufferView =
            java.nio.ByteBuffer.wrap(buffer.asByteArray(), byteOffset.toInt(), byteLength.toInt())
                .order(ByteOrder.BIG_ENDIAN)
    }

    fun getFloat32(index: Double, littleEndian: Boolean): Double {
        return if (littleEndian) {
            _littleEndianBufferView
        } else {
            _bigEndianBufferView
        }.getFloat((byteOffset + index).toInt()).toDouble()
    }


    fun getInt16(index: Double, littleEndian: Boolean): Double {
        return if (littleEndian) {
            _littleEndianBufferView
        } else {
            _bigEndianBufferView
        }.getShort((byteOffset + index).toInt()).toDouble()
    }
}
